lib/throttle-stream.js

```diff
@@ -0,0 +1,153 @@
+ * the desired bytes per second value.
+ *
+ * @param {Object} opts an options object or the "bps" function value
+ * @api public
+ * @constructor ThrottleStream
+ * @extends Transform
+ * @param {Object} [opts] - options
+ * @param {number|Function} [opts.bps] - number or function that returns bytes per second to throttle to
+ */
+function ThrottleStream (opts) {
+  if (!new.target) {
+    return new ThrottleStream(opts)
+  }
+
+  Transform.call(this, opts)
+
+  /**
+   * The function to calculate bytes per second to throttle to
+   * @type {Function}
+   * @public
+   * @memberof ThrottleStream
+   * @name bpsFn
+   * @default null
+   */
+  this.bpsFn = typeof opts?.bps === 'undefined'
+    ? function bpsFn (_elapsedTime, _bytes) {
+      return 16384
+    }
+    : typeof opts.bps === 'number'
+      ? function bpsFn (_elapsedTime, _bytes) {
+        return opts.bps
+      }
+      : opts.bps
+
+  /**
+   * The time that the stream started
+   * @type {number}
+   * @public
+   * @memberof ThrottleStream
+   * @name startTime
+   * @default null
+   */
+  this.startTime = null
+
+  /**
+   * The number of bytes that have been sent through the stream
+   * @type {number}
+   * @public
+   * @memberof ThrottleStream
+   * @name bytes
+   * @default 0
+   */
+  this.bytes = 0
+
+  /**
+   * The buffered chunk.
+   * @type {Buffer}
+   * @private
+   * @memberof ThrottleStream
+   * @name _buffer
+   * @default null
+   */
+  this._buffer = null
+
+  this._interval = null
+
+  /**
+   * Number of bytes allowed to be pushed to the next stream
+   * @type {number}
+   * @private
+   * @default 0
+   */
+  this._allowedBytes = 0
+
+  this._callback = null
+}
+
+ThrottleStream.prototype = Object.create(Transform.prototype, {
+  constructor: {
+    value: ThrottleStream,
+    enumerable: false,
+    writable: true,
+    configurable: true
+  }
+})
+
+/**
+ * The internal `_transform` function is called by the `write` function of the
+ * `Transform` stream base class. This function is not intended to be called
+ * directly.
+ * @param {Buffer} chunk
+ * @param {string} encoding
+ * @param {Function} fn
+ */
+ThrottleStream.prototype._transform = function _transform (chunk, encoding, callback) {
+  this.startTime === null && this._init()
+
+  chunk && (this._buffer = chunk)
+  callback && (this._callback = callback)
+
+  if (this._allowedBytes !== 0) {
+    const chunkLength = this._buffer.length
+    if (chunkLength <= this._allowedBytes) {
+      this._allowedBytes -= chunkLength
+      this.push(this._buffer)
+      this._buffer = null
+      this._callback()
+      return
+    }
+
+    const allowedBytes = this._allowedBytes
+    this._allowedBytes = 0
+    this.push(this._buffer.subarray(0, allowedBytes))
+    this._buffer = this._buffer.subarray(allowedBytes)
+  }
+}
+
+ThrottleStream.prototype._init = function _init () {
+  this.startTime = Date.now()
+
+  this._allowedBytes = this.bpsFn(0, 0)
+  this._interval = setInterval(() => {
```

Comment:
I wonder how test-friendly it is. If someone mocked out intervals in their code, this would end up in tests hanging if the plugin is used. It would be great if there was a fastify onClose hook for stopping all instances of this setInterval

Reply:
Wouldnt this mean that i kind of have to keep ALL streams in a big observable class just for the case I want to kill all of them?

Reply:
I don't know. Can you test it out with Jest timer mocking?

Reply:
I think I fixed it, by replacing the on 'end' listener with a _destroy

Reply:
Can you add a test for this?

Reply:
Maybe not today. I have to think how i can do this. Maybe a slow random stream and while it is streaming, emit an error event.

Reply:
Maybe we can introduce a stream implementation that throws an error?

Reply:
added unit test